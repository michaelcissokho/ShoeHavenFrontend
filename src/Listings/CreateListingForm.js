import React, { useState } from 'react'
import { useHistory } from 'react-router-dom'
import { Form, Container, Button } from 'react-bootstrap'

const CreateListingForm = ({ createListing }) => {
    const INITIAL_STATE = {
        title: '',
        picture: '',
        price: '',
        details: ''
    }

    const [formData, setFormData] = useState(INITIAL_STATE)

    let history = useHistory()

    const handleSubmit = (e) => {
        e.preventDefault()

        const { title, picture, price, details } = formData

        createListing(title, picture, price, details)

        history.push('/listings')

        setFormData(INITIAL_STATE)

    }

    const handleChange = (e) => {
        e.preventDefault()

        setFormData(formData => (
            {
                ...formData,
                [e.target.name]: e.target.value
            }
        ))
    }

    return (
        <div>
            <Container>
                <h1>New Listing:</h1>
                <Form onSubmit={handleSubmit}>
                    <Form.Group className="mb-3">
                        <Form.Label>Title</Form.Label>
                        <Form.Control
                            type="text"
                            name='title'
                            placeholder="Item Name"
                            value={formData.title}
                            onChange={handleChange}
                        />
                    </Form.Group>

                    <Form.Group className="mb-3">
                        <Form.Label>Picture</Form.Label>
                        <Form.Control
                            type="text"
                            name='picture'
                            placeholder="Link to Picture Of Item"
                            value={formData.picture}
                            onChange={handleChange}
                        />
                    </Form.Group>

                    <Form.Group className="mb-3">
                        <Form.Label>Price</Form.Label>
                        <Form.Control
                            type="number"
                            name='price'
                            placeholder="price"
                            value={formData.price}
                            onChange={handleChange}
                        />
                    </Form.Group>

                    <Form.Group className="mb-3">
                        <Form.Label>Details</Form.Label>
                        <Form.Control
                            as="textarea"
                            style={{height: '500px'}}
                            name='details'
                            placeholder="Description"
                            value={formData.details}
                            onChange={handleChange}
                        />
                    </Form.Group>

                    <Button variant="primary" type="submit">
                        Add New Lisiting
                    </Button>
                </Form>
            </Container>
        </div>
    )
}

export default CreateListingForm