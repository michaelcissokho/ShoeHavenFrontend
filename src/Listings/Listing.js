import React from 'react'
import { Card, Button, Row, Col } from 'react-bootstrap'

const Listing = ({ id, username, title, picture, price, details, addToCart, deleteListing }) => {

    return (
        <div>
            <Card style={{ width: '18rem' }}>
                <Card.Header> Seller: {username} </Card.Header>
                <Card.Img variant="top" src={picture} alt='Cool Listing' />
                <Card.Body>
                    <Card.Title>{title}</Card.Title>
                    <Card.Text>{details}</Card.Text>
                    <Card.Text>Price: ${price}</Card.Text>
                    <Row>
                        <Button as={Col} onClick={() => addToCart({ id, username, title, picture, price })} variant="success">
                            Add To Cart
                    </Button>
                        {(localStorage.getItem('username') === username) &&
                            <Button as={Col} onClick={() => deleteListing(id)}
                                variant='danger'>
                                Delete Listing
                        </Button>}
                    </Row>
                </Card.Body>
            </Card>
        </div>
    )
}


export default Listing