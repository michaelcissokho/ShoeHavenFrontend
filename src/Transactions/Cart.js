import React from 'react'
import CartItem from './CartItem'
import {Link} from 'react-router-dom'
import {v4 as uuid} from 'uuid'
import {Button} from 'react-bootstrap'

const Cart = ({items, removeFromCart}) => {
    return(
        <div>
            <h1>Cart:</h1>
            {items.map((item) => (
                          <CartItem 
                          key={uuid()} 
                          id={item.id}
                          username={item.username} 
                          title={item.title} 
                          picture={item.picture} 
                          price={item.price} 
                          removeFromCart={removeFromCart}
                          />
            ))}
            <br></br><br></br>
            <Button as={Link} to='/checkout' variant='success'>
                Checkout
            </Button>
        </div>
    )

}

export default Cart