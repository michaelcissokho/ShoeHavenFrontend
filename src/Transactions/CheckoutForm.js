import React, { useState } from 'react'
import { useHistory } from 'react-router-dom'
import { Container, Form, Row, Col, Button } from 'react-bootstrap'

const CheckoutForm = ({ cart, checkoutItems }) => {
    let total = 0
    for (let item of cart) {
        total += item.price
    }

    const history = useHistory()

    const INITIAL_STATE = {
        firstname: '',
        lastname: '',
        address: '',
        card: '',
    }

    const [formData, setFormData] = useState(INITIAL_STATE)

    function handleChange(e) {
        setFormData((formData) => (
            {
                ...formData,
                [e.target.name]: e.target.value
            }
        ))
    }

    function handleSubmit(e) {
        e.preventDefault()

        checkoutItems()

        setFormData(INITIAL_STATE)

        history.push('/listings')
    }

    return (
        <div>
            <h3>Cart Total: ${total}</h3>
            <Container>
                <h1>Enter Details To Checkout Your Items:</h1>
                <Form onSubmit={handleSubmit}>
                    <Row>
                        <Form.Group as={Col} className="mb-3">
                            <Form.Label>First Name</Form.Label>
                            <Form.Control
                                type="text"
                                name='firstname'
                                placeholder="firstname"
                                value={formData.firstname}
                                onChange={handleChange}
                            />
                        </Form.Group>

                        <Form.Group as={Col} className="mb-3">
                            <Form.Label>Last Name</Form.Label>
                            <Form.Control
                                type="text"
                                name='lastname'
                                placeholder="Last Name"
                                value={formData.lastname}
                                onChange={handleChange}
                            />
                        </Form.Group>
                    </Row>

                    <Form.Group className="mb-3">
                        <Form.Label>Address</Form.Label>
                        <Form.Control
                            type="text"
                            name='address'
                            placeholder="Address"
                            value={formData.address}
                            onChange={handleChange}
                        />
                    </Form.Group>

                    <Form.Group className="mb-3">
                        <Form.Label>Card Number</Form.Label>
                        <Form.Control
                            type="text"
                            name='card'
                            placeholder="Card Number"
                            value={formData.card}
                            onChange={handleChange}
                        />
                    </Form.Group>

                    <Button variant="success" type="submit">
                        Checkout
                    </Button>
                </Form>
            </Container>
        </div>
    )
}

export default CheckoutForm