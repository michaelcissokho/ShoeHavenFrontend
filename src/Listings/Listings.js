import React, { useEffect, useState } from 'react'
import Listing from './Listing'
import { ShoeHavenApi as api } from '../ShoeHavenApi'
import { v4 as uuid } from 'uuid'
import { Container, Row } from 'react-bootstrap'

const Listings = ({ addToCart, deleteListing }) => {
    const [listings, setListings] = useState([])
    
    useEffect(() => {
        async function pullListings() {
            let res = await api.request(`listings`)
            setListings(res.filter((item) => item.sold === false))
        }
        pullListings()
    }, [])

    return (
        <div>
            <h1>
                Listings:
            </h1>
            <Container>
                <Row xs={4}>
                    {listings.map((listing) =>
                        <Listing
                            key={uuid()}
                            id={listing.id}
                            details={listing.details}
                            picture={listing.picture}
                            price={listing.price}
                            title={listing.title}
                            username={listing.username}
                            addToCart={addToCart}
                            deleteListing={deleteListing}
                        />)}
                </Row>
            </Container>
        </div>
    )
}

export default Listings