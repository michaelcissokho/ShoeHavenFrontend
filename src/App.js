import React, { useState } from 'react'
import Home from './Home'
import User from './Users/User';
import Listings from './Listings/Listings'
import CreateListingForm from './Listings/CreateListingForm';
import NewPostForm from './Posts/NewPostForm'
import UpdateUserForm from './Users/UpdateUserForm'
import LoginPage from './Users/LoginPage'
import SignupPage from './Users/SignupPage'
import { Route, Switch } from 'react-router-dom'
import PostPage from './Posts/PostPage'
import Navigation from './Navigation'
import Cart from './Transactions/Cart'
import { ShoeHavenApi as api } from './ShoeHavenApi'
import CheckoutForm from './Transactions/CheckoutForm';
import 'bootstrap/dist/css/bootstrap.min.css';

function App() {
  //user auth or profile functions
  const [userLoggedIn, setUserLoggedIn] = useState(localStorage.getItem('username'))

  async function signup(username, password, firstname, lastname, email) {
    await api.signup(username, password, firstname, lastname, email)
    setUserLoggedIn(localStorage.getItem('username'))
  }

  async function login(username, password) {
    await api.login(username, password)
    setUserLoggedIn(localStorage.getItem('username'))
    return new Promise((resolve, reject) => {
      resolve()
    });
  }

  function logout() {
    localStorage.removeItem('username')
    localStorage.removeItem('token')
    alert('Logged Out')
    setUserLoggedIn(localStorage.getItem('username'))
  }

  async function updateUser(password, firstname, lastname, email) {
    await api.updateUser(password, firstname, lastname, email)
  }

  //listing functions
  async function createListing(title, picture, price, details) {
    await api.request(`listings/new`, { title, picture, price, details, sold: false }, 'post')
  }

  async function deleteListing(id) {
    await api.request(`listings/${id}`, {}, 'delete')

    alert('Listing Deleted')
  }

  //post functions
  async function createPost(title, body, picture) {
    await api.request(`posts/new`, { title, body, picture }, 'post')
  }

  async function deletePost(id) {
    await api.request(`posts/${id}`, {}, 'delete')

    alert('Post Deleted')
  }

  //cart functions
  const [cart, setCart] = useState([])

  function addToCart(cartItem) {
    setCart(cart => (
      [
        ...cart,
        cartItem
      ]
    ))

    alert('Added To Cart')
  }

  function removeFromCart(id) {
    setCart(
      cart.filter(cartItem => (cartItem['id'] !== id))
    )
  }
  async function checkoutItems() {
    for (let item of cart) {
      await api.createSale(item.id, item.username)
    }
    setCart([])
  }

  return (
    <div className="App">
        <Navigation logout={logout} userLoggedIn={userLoggedIn}/>
        <Switch>
          <Route exact path='/'>
            <Home userLoggedIn={userLoggedIn}/>
          </Route>
          <Route exact path='/login'>
            <LoginPage login={login} />
          </Route>

          <Route exact path='/signup'>
            <SignupPage signup={signup} />
          </Route>

          <Route exact path='/listings'>
            <Listings addToCart={addToCart} deleteListing={deleteListing} />
          </Route>

          <Route exact path='/posts'>
            <PostPage deletePost={deletePost} />
          </Route>

          <Route exact path='/posts/new'>
            <NewPostForm createPost={createPost} />
          </Route>

          <Route exact path='/users/:username'>
            <User />
          </Route>

          <Route exact path='/listings/new'>
            <CreateListingForm createListing={createListing} />
          </Route>

          <Route exact path='/:username/update'>
            <UpdateUserForm updateUser={updateUser} />
          </Route>

          <Route exact path='/cart'>
            <Cart items={cart} removeFromCart={removeFromCart} />
          </Route>

          <Route exact path='/checkout'>
            <CheckoutForm cart={cart} checkoutItems={checkoutItems} />
          </Route>

        </Switch>
    </div>
  );
}

export default App;
