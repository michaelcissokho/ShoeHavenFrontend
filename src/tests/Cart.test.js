import React from "react";
import '@testing-library/jest-dom'
import { render, fireEvent } from "@testing-library/react";
import Cart from "../Transactions/Cart";
import CheckoutForm from '../Transactions/CheckoutForm'
import App from '../App'
import Listings from '../Listings/Listings'
import { MemoryRouter, BrowserRouter } from 'react-router-dom'

it("renders without crashing", function () {
  render(<BrowserRouter><Cart items={[]} /></BrowserRouter>);
});

it("matches snapshot with no itms", function () {
  const { asFragment } = render(<BrowserRouter><Cart items={[]} /></BrowserRouter>);
  expect(asFragment()).toMatchSnapshot();
});


//click on a listing from listings page and see if it shows up in cart*********************** need help here
it('adds item to cart', function () {
  const { getByText } = render(
    <BrowserRouter>
      <Listings />
    </BrowserRouter>
  )
  const addToCart = getByText('Add To Cart')
  fireEvent.click(addToCart)
  //make sure you can close the alert*****************
  const closeAlert = getByText('OK')
  fireEvent.click(closeAlert)
  const cart = getByText('Cart')
  fireEvent.click(cart)
  expect(getByText(' Remove From Cart ')).toBeInTheDocument()

})


//check if you can delete an item from the cart
it('can delete item from cart', function () {
  //figure out how to mock function removeFrom cart**********
  const { getByText } = render(
    <BrowserRouter>
      <Cart items={[{ id: 1, username: 'test1', title: 'item1', picture: '', price: 100, removeFromCart: 'removeFromCart' }]} />)
  </BrowserRouter>
  )
  const link = getByText('Remove From Cart')
  fireEvent.click(link)
  expect(getByText('Remove From Cart')).not.toBeInTheDocument()
})


//click on checkout button and make sure it routes to checkoutform
it('goes to checkout form', function () {
  //figure out how to mock function removeFrom cart**********
  const { getByText } = render(<MemoryRouter initialEntries={['/cart']}><App /></MemoryRouter>)
  const link = getByText('Checkout')
  fireEvent.click(link)
  expect(getByText('Enter Details To Checkout Your Items:')).toBeInTheDocument()
})


//make sure checkout form renders properly
it('renders checkout form properly', function () {
  const { asFragment } = render(
    <BrowserRouter>
      <CheckoutForm cart={[{ id: 1, username: 'test1', title: 'item1', picture: '', price: 100, removeFromCart: 'removeFromCart' }]} />)
  </BrowserRouter>
  )
  expect(asFragment()).toMatchSnapshot()
})


