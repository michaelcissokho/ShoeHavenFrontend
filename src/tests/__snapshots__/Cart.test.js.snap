// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with no itms 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Cart:
    </h1>
    <br />
    <br />
    <a
      class="btn btn-success"
      href="/checkout"
      role="button"
      tabindex="0"
    >
      Checkout
    </a>
  </div>
</DocumentFragment>
`;

exports[`renders checkout form properly 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Cart Total: $100
    </h3>
    <div
      class="container"
    >
      <h1>
        Enter Details To Checkout Your Items:
      </h1>
      <form
        class=""
      >
        <div
          class="row"
        >
          <div
            class="mb-3 col"
          >
            <label
              class="form-label"
            >
              First Name
            </label>
            <input
              class="form-control"
              name="firstname"
              placeholder="firstname"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-3 col"
          >
            <label
              class="form-label"
            >
              Last Name
            </label>
            <input
              class="form-control"
              name="lastname"
              placeholder="Last Name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            Address
          </label>
          <input
            class="form-control"
            name="address"
            placeholder="Address"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            Card Number
          </label>
          <input
            class="form-control"
            name="card"
            placeholder="Card Number"
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-success"
          type="submit"
        >
          Checkout
        </button>
      </form>
    </div>
  </div>
  )
</DocumentFragment>
`;
